### PR TITLE
Ensure tasks are not garbage collected early

### PR DESCRIPTION
**Pull request by mikeshardmind**
## Summary

Adds a task store to `Client` and utilizes it in `_schedule_event`.

This fixes an issue identified in Discord here: https://discord.com/channels/336642139381301249/1115863485692256286
While it is unclear if this is the specific cause of the issue they experienced, it is possible to confirm that this is an issue.
A minimal synthetic example of this as an issue has been shared in Discord, and I'll include it as a comment below as well.

## Checklist

- [X] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [X] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by mikeshardmind** [comment]
Synthetic example of this as an issue:
```py

import gc
import asyncio
from typing import Coroutine, Any, Callable

class MinimalRepro:

    def __init__(self):
        self.loop = asyncio.get_running_loop()

    async def on_error(self, event_method: str, /, *args: Any, **kwargs: Any) -> None:
        # original implementation used `_log.exception`
    
        print('Ignoring exception in %s', event_method)

    # copied as-is
    async def _run_event(
        self,
        coro: Callable[..., Coroutine[Any, Any, Any]],
        event_name: str,
        *args: Any,
        **kwargs: Any,
    ) -> None:
        try:
            await coro(*args, **kwargs)
        except asyncio.CancelledError:
            pass
        except Exception:
            try:
                await self.on_error(event_name, *args, **kwargs)
            except asyncio.CancelledError:
                pass
    
    #copied as-is
    def _schedule_event(
        self,
        coro: Callable[..., Coroutine[Any, Any, Any]],
        event_name: str,
        *args: Any,
        **kwargs: Any,
    ) -> asyncio.Task:
        wrapped = self._run_event(coro, event_name, *args, **kwargs)
        # Schedules the task
        return self.loop.create_task(wrapped, name=f'discord.py: {event_name}')


async def useless():
    await asyncio.sleep(60)
    print("done being useless")

async def repro_case():
    m = MinimalRepro()
    m._schedule_event(useless, "useless")
    gc.collect()


if __name__ == '__main__':
    asyncio.run(repro_case(), debug=True)
```

File: discord/client.py [hunk 1]
```diff
@@ -289,6 +289,7 @@ def __init__(self, *, intents: Intents, **options: Any) -> None:
         self._application: Optional[AppInfo] = None
         self._connection._get_websocket = self._get_websocket
         self._connection._get_client = lambda: self
+        self.__task_store: set[asyncio.Task] = set()
```
Comment (Gobot1234) [inline discord/client.py:292]:
```suggestion
        self.__task_store: Set[asyncio.Task] = set()
```
Should use old classes for consistency right?

Comment (mikeshardmind) [reply discord/client.py]: Yeah, it should. This has been changed and an import added for it, my fault for forgetting that's not valid in 3.8

**Comment by mikeshardmind** [comment]
After further investigation (discussed in discord) I'm having trouble showing definitively that this can happen in an actual bot. There may be some implementation detail keeping tasks alive here, and I really don't want to dig into it, seems to not be the actual cause of issues here.

**Pull request closed without merging**